### PR TITLE
Proj fix

### PR DESCRIPTION
**Pull request by agramfort**
@Eric89GXL @dengemann if you have time to have a look.

The problem was raised in : https://github.com/agramfort/mne-python/commit/45883b048a8f249e3093fed74529f732f1bd163b

To avoid this local cooking and guarantee that proj attribute == True means all available proj have been applied I introduced a property. But now I broke the "delayed" option in epochs... The epochs code is really crazy now...

thanks for any help. It's a blocker for #1785

### CODE REVIEW

**Comment by Eric89GXL** [comment]
I think @dengemann is a better choice to look into this since he pushed the delayed projection code -- I haven't looked at it in depth so I don't fully understand it.

**Comment by agramfort** [comment]
travis happy !

reviewers be my guest

File: mne/io/proj.py [hunk 1]
```diff
@@ -101,11 +105,15 @@ def apply_proj(self):
...
         """
-        if self.info['projs'] is None:
+        if self.info['projs'] is None or len(self.info['projs']) == 0:
             logger.info('No projector specified for this dataset.'
                         'Please consider the method self.add_proj.')
             return self
 
+        # Exit delayed mode if you apply proj
+        if hasattr(self, '_delayed_proj'):
+            del self._delayed_proj
```
Comment (dengemann) [inline mne/io/proj.py]: add some verbosity here, something like:

logger.info('Leaving delayed SSP')

**Comment by dengemann** [comment]
@agramfort the interactive plots were working last time we looked together. Is that still the case?

**Comment by agramfort** [comment]
done and yes it's working

**Comment by agramfort** [comment]
let's wait for @Eric89GXL if you want to have a look

**Comment by agramfort** [comment]
@Eric89GXL any chance to have your feedback quickly?

**Comment by Eric89GXL** [comment]
Okay I still don't quite get it. It seems like we have three possible states:

1. Projectors have actually been applied to the data.

2. Projectors haven't been applied, but will be applied when it's necessary: delayed at the evoked stage, or delayed at the epochs stage, or at the inverse stage.

3. Projectors are present in the data, but will not be applied to the data at any stage.

Is this right? If so, how are these three states reflected in `raw.proj` / `epochs.proj` / `evoked.proj` attributes, and in `info['proj']`, especially related to "active" vs "inactive" proj states?

I think we need this clarified somewhere in the docstring once and for all.

**Comment by agramfort** [comment]
the idea is the following

- .proj attribute is now a property which is True if all projs are active. They might not be applied yet if data are not preloaded. In this case it's the _projector attribute that does the job.

- the proj parameter pass in constructor of raw or epochs calls apply_proj and hence after the .proj attribute is True

clear?

**Comment by Eric89GXL** [comment]
So if I use `proj=False` all the time, when will projectors be applied? When applying the inverse?

When will projectors be set to inactive? Never?

**Comment by agramfort** [comment]
> So if I use proj=False all the time, when will projectors be applied? When applying the inverse?

they should never be applied and .proj should be False.

> When will projectors be set to inactive? Never?

they will be inactive untill apply_proj() is called.

ok?

**Comment by Eric89GXL** [comment]
I want to be able to answer the following question, and I can't. What are the states of the following three variables:

1. `object.proj`

2. `object.info['projs']` `'active'` vs `'inactive'`

3. `object._data` having projectors applied to it or not

`for object [raw, epochs, evoked, cov]` when I do any combination of the following:

1. Construct `raw` then do `apply_proj` (or not)

2. After either of those, construct `Epochs` with `proj=True` or `proj=False`

3. After any of those, compute covariance (frow raw or epochs)

4. After any of these, apply an inverse operator to the data using any of these covs and any of these raw / epochs / evoked.

**Comment by Eric89GXL** [comment]
...and which of these is the suggested pipeline? Which of them is the "delayed SSP" set of operations? (Are those the same for us?)

**Comment by agramfort** [comment]
> I want to be able to answer the following question, and I can't. What are the states of the following three variables:
>
> object.proj

True if all projs are active. Else False.

> object.info['projs'] 'active' vs 'inactive'

they can be all active or inactive

> object._data having projectors applied to it or not

if object._data is there then the projs applied to it are the ones
marked as active.

> for object [raw, epochs, evoked, cov] when I do any combination of the following:
>
> Construct raw then do apply_proj (or not)
>
> After either of those, construct Epochs with proj=True or proj=False
>
> After any of those, compute covariance (frow raw or epochs)
>
> After any of these, apply an inverse operator to the data using any of these covs and any of these raw / epochs / evoked.

as soon as you've applied the projs it will stay active in the
remaining pipeline.

the suggested pipeline is proj=True in epochs (it's cheaper than for raw)

when you use delayed SSP projs are applied when you call get_data on
epochs. They are not applied to the evoked._data unless you call
apply_proj. The reason is that you want to reject with projs although
it's to stored in proj mode.

we can talk on hangout if you want.

**Comment by Eric89GXL** [comment]
That makes sense to me. Can you put a clear description in the docstring somewhere? My concern is that the next person who goes to deal with the `proj` code won't understand the relationships between all of these stages.

**Comment by agramfort** [comment]
pushed a docstring to the ProjMixin class.

**Comment by Eric89GXL** [comment]
LGTM, but you have angered the PEP8 gods

**Comment by Eric89GXL** [comment]
Oh wait nevermind, I think the error was unrelated. I restarted the Travis run. I'll merge once I see green. Thanks for the clarifications.

**Comment by agramfort** [comment]
travis is happy :)

**Comment by Eric89GXL** [comment]
thanks!

**Pull request merged**